File: 3.2.2/app/src/main/java/com/example/a33/MainActivity.java
package com.example.a33;

import android.os.Bundle;

import com.google.android.material.snackbar.Snackbar;

import androidx.appcompat.app.AppCompatActivity;

import android.view.View;

import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.example.a33.databinding.ActivityMainBinding;

import android.view.Menu;
import android.view.MenuItem;


import android.net.Uri;
import android.os.Bundle;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    WebView webView;
    String str;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        webView = findViewById(R.id.webView);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebChromeClient(new WebChromeClient());
        webView.setWebViewClient(new WebViewClient());

        if (getIntent().getExtras() == null) {
            str = "https://www.xe.com/";
        } else {
            Uri uri = getIntent().getData();
            if (uri != null) {
                str = uri.getQueryParameter("url");
            }
        }

        webView.loadUrl(str);
    }
}

File: 3.1/VulnerableApp/app/src/main/java/com/example/vulnerableapp/MainActivity.kt
package com.example.vulnerableapp

import android.net.Uri
import android.os.Bundle
import android.webkit.JavascriptInterface
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val webView: WebView = findViewById(R.id.webView)
        webView.webViewClient = WebViewClient()
        webView.settings.javaScriptEnabled = true




        webView.addJavascriptInterface(WebAppInterface(), "Android")


        val uri = intent.data
        if (uri != null) {
            webView.loadUrl(uri.toString())
        }
    }

    inner class WebAppInterface {
        @JavascriptInterface
        fun getFlag(): String {
            return "This is a sensitive flag!"
        }
    }
}
